File: solutions/csharp/binary-search/1/BinarySearch.cs
public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        var left = 0;
        var right = input.Length - 1;
        while (left < right)
        {
            var middle = (left + right + 1) / 2;
            if (input[middle] <= value)
            {
                left = middle;
            }
            else
            {
                right = middle - 1;
            }
        }
        return input[left] == value ? left : -1;
    }
}

File: solutions/haskell/binary-search/1/src/BinarySearch.hs
module BinarySearch (find) where

import Data.Array

f :: (Ord a) => Array Int a -> a -> Int -> Int -> Maybe Int
f arr x l r =
  if l == r
    then
      if arr ! l == x
        then
          Just l
        else
          Nothing
    else
      let m = div (l + r + 1) 2
       in if arr ! m <= x
            then
              f arr x m r
            else
              f arr x l (m - 1)

find :: (Ord a) => Array Int a -> a -> Maybe Int
find array x =
  if null array
    then
      Nothing
    else
      let left = 0
          right = length array - 1
       in f array x left right
